---
title: "Untitled"
output: html_document
date: "2025-04-20"
---

```{r message=F}
library(Biostrings)
library(BSgenome.Hsapiens.UCSC.hg38)
library(tidyplots)
library(GenomicRanges)
library(eulerr)
library(tidyverse)
```

```{r}
source('src.R')
```

### Boundary overlap between RCMC and Micro-C

```{r}
celltypes = c('GM12878', 'HCT116', 'K562', 'H1')
```

```{r}
# read in compartments generated by call_insulation_domains.py
all_rcmc_cmpts = set_names(celltypes) |> 
    map(~ import(paste0('compartment_analyses/rcmc_compartments/', .x, '_compartments_20kb.bed'))) |> 
    map(~ as_tibble(.x)) |> 
    map(~ mutate(.x, id = paste(seqnames, start, end, sep = '_')))

all_microc_cmpts = set_names(celltypes) |> 
    map(~ import(paste0('compartment_analyses/microc_res_200_missing_filtered/', .x, '_compartments_20kb.bed'))) |> 
    map(~ subsetByOverlaps(.x, capture_regions)) |> 
    map(~ as_tibble(.x)) |> 
    map(~ mutate(.x, id = paste(seqnames, start, end, sep = '_')))
```

```{r}
plot_list = list()
for (celltype in celltypes){
    euler_list = list('RCMC' = all_rcmc_cmpts[[celltype]]$id,
                   'Micro-C' = all_microc_cmpts[[celltype]]$id)
        
    plot_list[[celltype]] = plot(euler(euler_list, shape = 'ellipse'), quantities = TRUE, fill =  c("#8B4500", "#CDB38B"),
                                 main = celltype)
}

wrap_plots(plot_list, nrow = 1)

ggsave('figures/cmpt_boundary_microc_rcmc_overlap.pdf', width = 11, height = 4)
```

### Downsampled boundary counts

```{r}
# read in compartments generated by call_insulation_domains.py
downsampled_compartments = fs::dir_ls(path = 'compartment_analyses/downsampled_compartments/', glob = '*compartments_20kb.bed') %>%
    as.character() %>%
    purrr::set_names(function(x) {
    str_sub(fs::path_file(x), end = - str_length('_compartments_20kb.bed') - 1)
    }) |> 
    map(~ import(.x))
```

```{r}
names(downsampled_compartments) |> 
    map(~ mutate(downsampled_compartments[[.x]], frac = .x)) |> 
    map(~ as_tibble(.x)) |> 
    reduce(bind_rows) |> 
    separate_wider_delim(frac, delim = '_', names = c(NA, NA, 'frac')) |> 
    bind_rows(mutate(all_rcmc_cmpts$GM12878, frac = '1')) |> 
    filter(frac != '0.00390625') |> 
    tidyplot(x = frac) |> 
    add_count_bar() |> 
    reverse_x_axis_labels() |> 
    adjust_colors(new_colors = 'black')

ggsave('figures/downsample_number_boudaries.pdf')
```

### Boundary overlap between celltypes upset plot

```{r}
read_boundaries = function(filename){
    read_tsv(filename,
         show_col_types = FALSE) |> 
    mutate(id = paste(chrom, start, end, sep = '_')) 
}
```

```{r}
boundaries = set_names(celltypes) |> 
    map(~ read_boundaries(paste0('compartment_analyses/rcmc_compartments/', .x, '_boundaries_20kb.bed')))
```

```{r}
lt = list(GM12878 = boundaries$GM12878$id,
     HCT116 = boundaries$HCT116$id,
     H1 = boundaries$H1$id,
     K562 = boundaries$K562$id
)
```

```{r}
pdf('figures/boundary_overlap_between_celltypes.pdf', width = 5, height = 3)
m1 = make_comb_mat(lt)
UpSet(m1)
dev.off()
```

### Write boundary sequences to fastq files for motif analyses

```{r}
GM12878_unique_boundaries = extract_comb(m1, comb_name = '1000')
HCT116_unique_boundaries = extract_comb(m1, comb_name = '0100')
H1_unique_boundaries = extract_comb(m1, comb_name = '0010')
K562_unique_boundaries = extract_comb(m1, comb_name = '0001')
```

```{r}
shared_boundaries =
    c('1111', '0101', '0110', '1010') |> 
    map(~ extract_comb(m1, comb_na = .x)) |> 
    reduce(c) |>
    as_tibble() |>
    separate_wider_delim(value, names = c('chrom', 'start', 'end'), delim = '_', cols_remove = FALSE) |>
    makeGRangesFromDataFrame(keep.extra.columns = TRUE)
```

```{r}
write_seqs(shared_boundaries, 'compartment_analyses/rcmc_unique_boundaries/shared_boundaries.fa')
```

```{r}
unique_boundaries = list('GM12878' = GM12878_unique_boundaries,
                         'HCT116' = HCT116_unique_boundaries,
                         'H1' = H1_unique_boundaries,
                         'K562' = K562_unique_boundaries) |> 
    map(~ as_tibble(.x)) |> 
    map(~ separate_wider_delim(.x, value, names = c('chrom', 'start', 'end'), delim = '_', cols_remove = FALSE)) 
    
unique_boundaries_gr = map(~ makeGRangesFromDataFrame(.x, keep.extra.columns = TRUE)) 
    
celltypes |> 
    map(~ write_seqs(unique_boundaries_gr[[.x]], paste0('compartment_analyses/', .x, '_unique_boundaries.fa')))
```

### Non compartment anchors for negative set

```{r}
capture_regions_200bp_tiles_no_cmpt_anchors = subsetByOverlaps(capture_regions_200bp_tiles, flank(reduce(unique_boundaries, c), 10000), invert = TRUE)
```

```{r}
set.seed(1018)
random_cmpt_neg_200bp_tiles = sample(capture_regions_200bp_tiles_no_cmpt_anchors, replace=FALSE, size = 1000)
write_seqs(random_cmpt_neg_200bp_tiles, 'compartment_analyses/rcmc_unique_boundaries/random_negative_200bp_anchors.fa')
```

### Boundary strength comparisons

```{r}
boundaries_shared_annotated = boundaries |> 
    map(~ left_join(.x, as_tibble(shared_boundaries) |> 
                    mutate(shared = 'shared') |> 
                    select(value, shared),
                    by = c('id' = 'value'))) 

celltypes |> 
    map(~ mutate(boundaries_shared_annotated[[.x]], celltype = .x)) |> 
    reduce(bind_rows) |> 
    mutate(shared = replace_na(shared, 'unique')) |> 
    ggplot(aes(log2_insulation_score_20000, col = shared)) +
    geom_density() + 
    plot_theme_facet() +
    facet_wrap(~ celltype, nrow = 1) +
    scale_color_manual(values = c('#EFD5B8', '#97735B')) +
    xlab('log2(insulation_score)') +
    ylab('density') +
    theme(legend.position = 'None')

ggsave('figures/shared_unique_cmpt_boundaries.pdf', width = 12, height = 4)
```