---
title: "Untitled"
output: html_document
date: "2025-04-21"
---

```{r}
library(GenomicRanges)
library(tidyverse)
library(AnnotationHub)
```

This file deals with pcHi-C loops from Mifsud et al., but can be applied to datasets from any other papers

### Read and filter Mifsud genes for RCMC regions only

Get GENCODE gene annotations

```{r}
ah = AnnotationHub()
# ensembl transcripts that corresponds to GENCODE v29, since that's what ENCODE used
edb = ah[["AH64923"]]
```

```{r}
gencode_v29_promoters = promoters(edb, upstream = 1000, downstream = 500, columns = c('gene_id', 'gene_id_version', 'gene_name'), filter = ~tx_biotype == 'protein_coding' & tx_support_level == 1)
seqlevelsStyle(gencode_v29_promoters) = 'UCSC'
```

GM12878 RCMC loops

```{r}
GM12878_loops = read_tsv('example_data/loop_examples/GM12878_rcmc_all_1kb.bedpe',
         col_names = c('seqnames1', 'start1', 'end1', 'seqnames2', 'start2', 'end2', 'name'))
```

```{r}
GM12878_loop_anchors = GM12878_loops |> 
    select(seqnames1, start1, end1) |>
    rename_with(~ sub("\\d+$", "", .x)) |>
    bind_rows(as_tibble(GM12878_loops) |>
    select(seqnames2, start2, end2) |> rename_with(~ sub("\\d+$", "", .x))) |>
    distinct() |> 
    makeGRangesFromDataFrame()
```

Find only loops that overlap with promoter annotations 

```{r}
GM12878_loop_anchor_gencode_overlap = findOverlaps(GM12878_loop_anchors, gencode_v29_promoters)

GM12878_loop_anchor_gencode_overlap =
    as_tibble(GM12878_loop_anchors[queryHits(GM12878_loop_anchor_gencode_overlap)]) |> 
    dplyr::select(-width, -strand) |> 
    bind_cols(as_tibble(gencode_v29_promoters[subjectHits(GM12878_loop_anchor_gencode_overlap)]) |> 
              dplyr::select(gene_id, gene_id_version, gene_name))
```

Find loops in pcHiC that overlap RCMC looped genes 

```{r}
# get loops that are in RCMC regions (hg38)
mifsud_loops_in_rcmc_regions =
    read_tsv('promoter_cap_HiC_comparisons/GM12878_signif_interactions_within_rcmc.bedpe', show_col_types = FALSE,
             col_names = c('chr1', 'start1', 'end1', 'chr2', 'start2', 'end2', 'Symbol', 'mifsud_loop_id')) |>
    select(-contains('X'))

# get interaction scores from original pcHiC file for loops (hg19) in RCMC regions
mifsud_interaction_scores = read_tsv('TS5_GM12878_promoter-other_significant_interactions.txt', show_col_types = FALSE) |>  
    mutate(mid1 = sprintf((`start bait` + `end bait`)/2, fmt = '%#.1f'),
           mid2 = sprintf((start + end)/2, fmt = '%#.1f')) |>
    mutate(mifsud_loop_id = paste(`chr bait`, mid1, mid2, sep = '_')) |>
    filter(`chr bait` == chr, mifsud_loop_id %in% mifsud_loops_in_rcmc_regions$mifsud_loop_id) 
```

```{r}
# get pcHiC loops that overlap RCMC genes
mifsud_genes_in_GM12878_rcmc = list()

for (i in seq_len(nrow(GM12878_loops_per_gencode))) {
    
    gene_name = GM12878_loops_per_gencode[i, ]$gene_name
    filtered_df = mifsud_interaction_scores |>
        dplyr::filter(grepl(gene_name, Symbol)) |> 
        select(mifsud_loop_id)

    mifsud_genes_in_GM12878_rcmc[[ensym(gene_name)]] = filtered_df
    # }
}
```

```{r}
# Get pcHi-C loops that are shared/unique (rcmc_merged_vs_capHiC.bedpe file from bedtools pairtopair)
mifsud_loops_overlap_anno =
    read_tsv('promoter_cap_HiC_comparisons/rcmc_merged_vs_capHiC.bedpe', col_names = FALSE, show_col_types = FALSE) |> 
    rename('loop_id' = 'X7', 'mifsud_loop_id' = 'X15') |>
    select(-contains('X')) |> 
    full_join(reduce(mifsud_genes_in_GM12878_rcmc, bind_rows)) |> 
    mutate(overlap = case_when(
        is.na(loop_id) ~ 'pcHiC_unique',
        TRUE ~ 'RCMC_overlap'
    )) 

# write to file for loop pileup
mifsud_loops_overlap_anno |>
    distinct(mifsud_loop_id, overlap) |>
    write_tsv('promoter_cap_HiC_comparisons/mifsud_promoter_capture_HiC_unique_loops.txt')
```

```{r}
# Get RCMC loops that are shared/unique (rcmc_merged_vs_capHiC.bedpe file from bedtools pairtopair)
read_tsv('promoter_cap_HiC_comparisons/rcmc_merged_vs_capHiC.bedpe', col_names = FALSE, show_col_types = FALSE) |> 
    rename('loop_id' = 'X7', 'mifsud_loop_id' = 'X15') |>
    select(-contains('X')) |>
    full_join(rcmc_loops_overlap_mifsud_genes) |> 
    mutate(overlap = case_when(
        is.na(mifsud_loop_id) ~ 'RCMC_unique',
        TRUE ~ 'RCMC_overlap'
    )) |>
    select(-mifsud_loop_id) |>
   write_tsv('promoter_cap_HiC_comparisons/GM12878_rcmc_unique_loops_vs_mifsud.txt')
```

### Plot loop strength comparisons

```{r}
GM12878_loop_strength = read_tsv('../example_data/loop_examples/GM12878_rcmc_loop_strength_oe_3kb.tsv', show_col_types = FALSE)
```

```{r}
# shared loops from rcmc_merged_vs_capHiC.bedpe file generated by bedtools pairtopair
read_tsv('promoter_cap_HiC_comparisons/rcmc_merged_vs_capHiC.bedpe', col_names = FALSE, show_col_types = FALSE) |> 
    rename('loop_id' = 'X7', 'mifsud_loop_id' = 'X15') |> 
    select(-contains('X')) |> 
    left_join(GM12878_loop_strength) |> 
    inner_join(mifsud_interaction_scores) |>
    filter(loop_strength > 0) |> 
    ggplot(aes(log2(loop_strength), `log(observed/expected)`)) +
    geom_point() +
    plot_theme() +
    stat_cor(method = 'pearson') 

ggsave('figures/pcHiC_loop_strength_vs_rcmc.pdf', width = 6, height = 6)
```

